ddtrace/internal/telemetry/__init__.py

```diff
@@ -0,0 +1,69 @@
+"""
+Instrumentation Telemetry API.
+This is normally started automatically by ``ddtrace-run`` when the
+``DD_INSTRUMENTATION_TELEMETRY_ENABLED`` variable is set.
+To start the service manually, invoke the ``enable`` method::
+    from ddtrace.internal import telemetry
+    telemetry.enable()
+"""
+
+from ...internal import forksafe
+from ..service import ServiceStatus
+from .writer import TelemetryWriter
+
+
+TELEMETRY_WRITER = TelemetryWriter()
+
+__all__ = ["enable", "disable", "add_integration"]
+
+
+def add_integration(integration_name, auto_enabled):
+    # type: (str, bool) -> None
+    """
+    Creates and queues the names and settings of a patched module
+
+    :param str integration_name: name of patched module
+    :param bool auto_enabled: True if module is enabled in _monkey.PATCH_MODULES
+    """
+    TELEMETRY_WRITER.add_integration(integration_name, auto_enabled)
+
+
+def _restart():
+    # type: () -> None
+    if TELEMETRY_WRITER.status == ServiceStatus.RUNNING:
+        disable()
+    enable()
+
+
+def disable():
+    # type: () -> None
+    """
+    Disable the telemetry collection service.
+    Once disabled, telemetry collection can be re-enabled by calling ``enable`` again.
+    """
+    if TELEMETRY_WRITER.status == ServiceStatus.STOPPED:
+        return
+
+    forksafe.unregister(_restart)
+
+    TELEMETRY_WRITER.stop()
+    TelemetryWriter.enabled = False
+
+    TELEMETRY_WRITER.join()
+
+
+def enable():
+    # type: () -> None
+    """
+    Enable the instrumentation telemetry collection service. If the service has already been
+    activated before, this method does nothing. Use ``disable`` to turn off the telemetry collection service.
+    """
+    if TELEMETRY_WRITER.status == ServiceStatus.RUNNING:
+        return
+
+    TELEMETRY_WRITER.start()
+    TelemetryWriter.enabled = True
+
+    forksafe.register(_restart)
+
+    TELEMETRY_WRITER.app_started_event()
```

Comment:
What if enable is called twice? Is this ok to be considered two app started events?

Reply:
It's not a huge deal. It would just duplicate the data in the backend

Reply:
would it be considered a duplicate since the timestamps are different?

or does it merge based on other app level information like runtime-id, etc?

Reply:
The runtime id is used as a unique identifier for application instances. We used to merge values based on runtime id (using resources platform). Since internal analytics does not support updates only writes (telemetry intake writes data to a csv which is then uploaded to an athena database via a cron job). It's very messy but duplicate data is currently filtered out using custom queries in metabase dashboards. This is honestly a major issue which needs to be address in future iterations. I don't think metabase is a great datastore for telemetry

Reply:
The runtime id is used as a unique identifier for application instances. We used to merge values based on runtime id but that code path is deprecated. Since internal analytics does not support updates only writes every payload we send will be stored. It's very messy but duplicate data is currently filtered out using custom queries in internal dashboards. This is honestly a major issue which needs to be address in future iterations